api/v1/calico_node_types.go

```diff
@@ -0,0 +1,108 @@
+}
+
+// CalicoNodeInitContainer is a calico-node DaemonSet init container.
+type CalicoNodeInitContainer struct {
+	// Name is an enum which identifies the calico-node DaemonSet init container by name.
+	// +kubebuilder:validation:Enum=install-cni;hostpath-init;flexvol-driver;mount-bpffs;node-certs-key-cert-provisioner;calico-node-prometheus-server-tls-key-cert-provisioner
+	Name string `json:"name"`
+	// Resources allows customization of limits and requests for compute resources such as cpu and memory.
+	Resources *v1.ResourceRequirements `json:"resources"`
+}
+
+// CalicoNodeDaemonSetPodSpec is the calico-node DaemonSet's PodSpec.
+type CalicoNodeDaemonSetPodSpec struct {
+	// InitContainers is a list of calico-node init containers.
+	// +optional
+	InitContainers []CalicoNodeInitContainer `json:"initContainers,omitempty"`
+	// Containers is a list of calico-node containers.
+	// +optional
+	Containers []CalicoNodeContainer `json:"containers,omitempty"`
+	// Affinity is a group of affinity scheduling rules for the calico-node pods.
+	// +optional
+	Affinity *v1.Affinity `json:"affinity,omitempty"`
+	// NodeSelector is the calico-node pod's scheduling constraints.
+	// +optional
+	NodeSelector map[string]string `json:"nodeSelector,omitempty"`
+	// Tolerations is the calico-node pod's tolerations.
+	// +optional
+	Tolerations []v1.Toleration `json:"tolerations,omitempty"`
+}
+
+// CalicoNodeDaemonSetPodTemplateSpec is the calico-node DaemonSet's PodTemplateSpec
+type CalicoNodeDaemonSetPodTemplateSpec struct {
+	// Metadata is a subset of a Kubernetes object's metadata that is added to
+	// the pod's metadata.
+	// +optional
+	Metadata *Metadata `json:"metadata,omitempty"`
+	// Spec is the calico-node DaemonSet's PodSpec.
+	// +optional
+	Spec *CalicoNodeDaemonSetPodSpec `json:"spec,omitempty"`
+}
+
+// CalicoNodeDaemonSet is the configuration for the calico-node DaemonSet.
+type CalicoNodeDaemonSet struct {
+	// Metadata is a subset of a Kubernetes object's metadata that is added to the DaemonSet.
+	// +optional
+	Metadata *Metadata `json:"metadata,omitempty"`
+	// Spec is the specification of the calico-node DaemonSet.
+	// +optional
+	Spec *CalicoNodeDaemonSetSpec `json:"spec,omitempty"`
+}
+
+// CalicoNodeDaemonSetSpec defines configuration for the calico-node DaemonSet.
+type CalicoNodeDaemonSetSpec struct {
+	// MinReadySeconds is the minimum number of seconds for which a newly created DaemonSet pod should
+	// be ready without any of its container crashing, for it to be considered
+	// available. Defaults to 0 (pod will be considered available as soon as it
+	// is ready).
+	// +optional
+	// +kubebuilder:validation:Minimum=0
+	// +kubebuilder:validation:Maximum=2147483647
+	MinReadySeconds *int32 `json:"minReadySeconds,omitempty"`
+	// Template describes the calico-node DaemonSet pod that will be created.
+	// +optional
+	Template *CalicoNodeDaemonSetPodTemplateSpec `json:"template,omitempty"`
+}
```

Comment:
With these nested struct pointers, I'm thinking it might be nice to have accessors to reach into the nested struct. E.g.:

```
// Container is an arbitrary container.
type Container struct {
	// Name is identifies the container by name.
	Name string `json:"name"`
	// Resources allows customization of limits and requests for compute resources such as cpu and memory.
	Resources *v1.ResourceRequirements `json:"resources"`
}

func (c *CalicoNodeDaemonSet) GetPodTemplateMetadata() *Metadata {
	if c.Spec != nil {
		if c.Spec.Template != nil {
			return c.Spec.Template.Metadata
		}
	}
	return nil
}

func (c *CalicoNodeDaemonSet) GetInitContainers() []Container {
	if c.Spec != nil {
		if c.Spec.Template != nil {
			if c.Spec.Template.Spec != nil {
				if c.Spec.Template.Spec.InitContainers != nil {
					cs := make([]Container, len(c.Spec.Template.Spec.InitContainers))
					for i, v := range c.Spec.Template.Spec.InitContainers {
						c := Container{Name: v.Name, Resources: v.Resources}
						cs[i] = c
					}
					return cs
				}
			}
		}
	}
	return nil
}
```